### ISSUE TITLE
getSnapshot() non-cached option

### ISSUE THREAD

**Issue by tsightler** [dgreif/ring#596]
# Feature Request

### Description
Hey Dusty.

Due to popular demand, I've recently implemented snapshot support in ring-mqtt so that users can get snapshots either on motion events, on an interval, or both.  The default behavior for interval snapshots is to get a new snapshot ever 30 seconds, for wired cameras, or ever 10 minutes for battery camera, although users can control the interval if they really want to.  This works, well, but of course happens continuously, even if someone is not viewing, because, since I'm using MQTT, I have no way to know if snapshot images are being viewed actively or not.

The issue comes with the feature which attempts to get an updated snapshot during a motion event.  The problem is, if there happens to be a motion event within 10 seconds of the most recent interval based snapshot, ring-client-api returns the cached snapshot.  With 30 second refreshes there's a 1 in 3 chance the image returned is the cached image instead of the motion image.

I fully understand the issues around battery cameras not being able to take snapshots on while recording a motion event.  I've worked around this as best I can by grabbing a still image from the live stream for battery cameras.  However, for powered cameras, I don't think there's any issue requesting a snapshot at any time, it's just the getSnapshot() implementation that returns a cached snapshot if called within 10 seconds.  In ring-mqtt I've implemented a workaround where, if the camera has an active ding event, I just call the Ring Rest APIs directly to force a new snapshot, effectively bypassing getSnapshot().  This seems to work well and isn't difficult at all, but seems less than ideal.

### Describe the solution you'd like
I was wondering if you might consider a small change to getSnapshot() so that a new ding event immediately invalidates the snapshot cache so the next call to getSnapshot() would always return an updated image regardless of when the last snapshot was taken.  Looking at the code I think it would be fairly easy to implement (although I'm not sure as it also looks like you might just pull the latest snapshot from the API if the timestamp is less than the expire time, but you could also condition this on active ding for example).   Implemented carefully I don't believe it would open up the door for any API abuse.

### Describe alternatives you've considered
I considered calling getSnapshot() on a loop until it returns a new image, but this can sometimes lead to a significant delay in snapshot delivery and, in some cases, whatever cause the motion event is already gone by then, so it limits the functionality somewhat.  I also considered always using the livestream to grab a snapshot like I do for battery cameras, but it feels heavy handed, isn't 100% reliable (I have error checks for cases where the stream fails to start and retries) and the quality of the image is variable (I try to grab a keyframe directly from the AVC stream, but it sometimes still has motion blur, maybe I still don't have my ffmpeg options just right).

**Comment by dgreif** [comment]
@tsightler glad to see you are still so active on the MQTT project! I think your understanding of the snapshot implementation is spot on.  I'll give it some thought on my end and see if there is a solution that works well for both homebridge and MQTT.  At first pass, I think the solution you proposed makes sense.

Alternatively, I think the _ideal_ solution for all parties would be if I can finally get tied into the push notifications interface directly from Ring.  They have a way to register web hooks, allowing clients to register for all ding events.  The info pushed in those scenarios likely contains the url for the snapshot captured from the motion/ding, which is used when showing rich notifications from the Ring app.  The problem is that Ring has this hook api restricted so that it can only be registered by certain types of clients (like iOS/Andriod/Windows official push APIs).  It's been a while since I reached out to Ring, so I'll ping them today and see if they are ready to work on getting this client integrated into that endpoint.

**Comment by tsightler** [comment]
Yep, the project continues to add users and more features are constantly requested, so I try to keep it running as much as possible, although it really needs another rewrite now that I'm at least approaching amateur level with Javascript, but time is a very limiting factor.

I agree 100% that getting the push notifications would be the best, I looked at that a little myself but didn't see any way to figure out where the images were so gave up quickly.  I was hoping they were just adding it to the existing snapshots but it seems clearly not.  This approach would certainly be ideal for battery cameras as grabbing a frame from the livestream, while seeming to work pretty well, opened up some other side effects I did not anticipate, like terminating the saved recording if I don't let the livestream run even after I get my single frame.  I guess it makes sense, but I never realized the recording length was somehow tied to whether the livestream was opened/answered or not.

I'll certainly take your lead here and just use whatever you eventually come up with.  Thanks for considering the request.
